linfa-clustering/src/k_means/algorithm.rs

```diff
@@ -162,31 +164,50 @@ impl<'a, F: Float, R: Rng + Clone, D: Data<Elem = F>, T: Targets> Fit<'a, ArrayB
         let mut rng = self.rng();
         let observations = dataset.records().view();
 
-        let mut centroids = get_random_centroids(self.n_clusters(), &observations, &mut rng);
-
-        let mut has_converged;
-        let mut n_iterations = 0;
-
+        let mut min_inertia = F::infinity();
+        let mut best_centroids = None;
+        let mut best_iter = None;
         let mut memberships = Array1::zeros(observations.dim().0);
 
-        loop {
-            update_cluster_memberships(&centroids, &observations, &mut memberships);
-            let new_centroids = compute_centroids(self.n_clusters(), &observations, &memberships);
-
-            let distance = centroids
-                .sq_l2_dist(&new_centroids)
-                .expect("Failed to compute distance");
-            has_converged = distance < self.tolerance() || n_iterations > self.max_n_iterations();
-
-            centroids = new_centroids;
-            n_iterations += 1;
-
-            if has_converged {
-                break;
+        let n_init = self.n_init();
+
+        for _ in 0..n_init {
+            let mut inertia = min_inertia;
+            let mut centroids = get_random_centroids(self.n_clusters(), &observations, &mut rng);
+            let mut converged_iter: Option<u64> = None;
+            for n_iter in 0..self.max_n_iterations() {
+                update_cluster_memberships(&centroids, &observations, &mut memberships);
+                let new_centroids =
+                    compute_centroids(self.n_clusters(), &observations, &memberships);
+                inertia = compute_inertia(&new_centroids, &observations, &memberships);
+                let distance = centroids
+                    .sq_l2_dist(&new_centroids)
+                    .expect("Failed to compute distance");
+                centroids = new_centroids;
+                if distance < self.tolerance() {
+                    converged_iter = Some(n_iter);
+                    break;
+                }
+            }
+            if inertia < min_inertia {
```

Comment:
can you add a reference where inertia is coming from? The `inertia` is just the total sum of error for each cluster assignment?

Reply:
I have added some comments: we keep the centroids which minimize the inertia (sum of squared distances of the closest centroid for all observations) over the n runs